### PR TITLE
Enhancements for first-paragraph date parsing

### LINKED ISSUE #77: Errors parsing birth and death dates from the first sentence
Richard Shakespeake was born in 1490 so he can't possibly still be alive, yet in Python WikiBase:

(get "wikibase-person" "Richard Shakespeare" (:code "BIRTH-DATE"))
((:YYYYMMDD 14900000))
(get "wikibase-person" "Richard Shakespeare" (:code "DEATH-DATE"))
((:ERROR ATTRIBUTE-VALUE-NOT-FOUND :REPLY "Currently alive"))

Compare Ruby WikiBase:

(get "wikipedia-person" "Richard Shakespeare" (:code "BIRTH-DATE"))
((:YYYYMMDD 14900000))
(get "wikipedia-person" "Richard Shakespeare" (:code "DEATH-DATE"))
((:YYYYMMDD 15610000))

Richard Shakespeare's first sentence says "Richard Shakespeare (1490–d. before 10 February 1561) was a husbandman of [snip]".  The death date is hard to parse, but python WikiBase should something reasonable, which could be:

  - return 1561 like ruby WikiBase
  - return a non-yyyymmdd value: ((:html "d. before 10 February 1561"))
  - deprecated option: return the empty set of values: ()
  - return the value both ways: ((:html "d. before 10 February 1561" :yyyymmdd 15610210))

### LINKED ISSUE #87: Return :yyyymmdd and :html for wikibase-person
[Originally raised by @sfelshin over email]

wikipedia-person BIRTH-DATE should be less quick to return :yyyymmdd.

Note that both classes return the same result:

$ telnet nauru 8023
(get "wikipedia-person" "Sarah Bernhardt" (:code "BIRTH-DATE"))
((:yyyymmdd 18441022))

$ telnet nauru 8023
(get "wikibase-person" "Sarah Bernhardt" (:code "BIRTH-DATE"))
((:yyyymmdd 18441022))

But if we look at the Wikipedia article https://en.wikipedia.org/wiki/Sarah_Bernhardt we find

> Sarah Bernhardt (French: [sa.ʁa bɛʁ.nɑʁt];[2] c. 22/23 October 1844 – 26 March 1923) was a French stage and early film actress.

and
  
> Born	Rosine Bernardt
> c. 22 October 1844[1]
> Paris, France

It seems to me that this would be a good opportunity to return something along the lines of

$ telnet nauru 8023
(get "wikipedia-person" "Sarah Bernhardt" (:code "BIRTH-DATE"))
```
((:yyyymmdd 18441022 :html "<tr>
<th scope="row">Born</th>
<td><span class="nickname">Rosine Bernardt</span><br />
c. <span style="display:none">(<span class="bday">1844-10-22</span>)</span>22 October 1844<sup id="cite_ref-1" class="reference"><a href="#cite_note-1"><span>[</span>1<span>]</span></a></sup><br />
<span class="birthplace">Paris, France</span></td>
</tr>"))
```

$ telnet nauru 8023
(get "wikibase-person" "Sarah Bernhardt" (:code "BIRTH-DATE"))
((:yyyymmdd 18441022 :html "c. 22/23 October 1844"))

### LINKED ISSUE #92: Separate wikibase-person BIRTH-DATE from BIRTH-DATE from infobox
@alvaromorales says:

> `wikibase-person`'s `BIRTH-DATE` often comes from the infobox. This functionality is described in the Ruby-WikipediaBase wiki page:

>> `BIRTH-DATE`, `wikibase-person`
>> Fetched from the infobox, or, if it is not found, from the article, or, if it is not found, the category information of the article. Always relies on the first date of birth found, matching one of several supported formats.

> (except for the bit about categories; we're not currently using them for dates)

> Perhaps it would make more sense to change the code and separate these implementations:

>   - `wikipedia-`*class* ==> date from the infobox
>   - `wikibase-person` ==> date from the first sentence of the article

> START would have the choice of which date to ask for. `wikibase-person` should probably be a fallback if the infobox doesn't contain a date attribute.

I would definitely prefer this.

### LINKED ISSUE #94: First-sentence date parsing should prefer precision to recall
Currently Beethoven appears to have been born in the year 17:

    (get "wikibase-person" "Ludwig van Beethoven" (:CODE "BIRTH-DATE"))
    ((:yyyymmdd 00170000))

The first sentence for our dear friend is:

> Ludwig van Beethoven (Listeni/ˈlʊdvɪɡ væn ˈbeɪˌtoʊvən/; German: [ˈluːtvɪç fan ˈbeːtˌhoˑfn̩] ( listen); baptised 17 December 1770[1] – 26 March 1827)

Admittedly that's hard to parse, but if it's too hard, then `wikibase-person` should give up and return the date in `:html` format rather than `:yyyymmdd`, or it should give up entirely and throw some sort of error.  It should not claim that Beethoven could have had Christ for a baby sitter.

### LINKED ISSUE #215: wikibase-person BIRTH-DATE returns name, not date
```
$ telnet kiribati 8023
(get "wikibase-person" "Marvin Minsky" (:CODE "BIRTH-DATE"))
((:html "Marvin Lee Minsky"))
```

DEATH-DATE, on the other hand, works just fine:
```
$ telnet kiribati 8023
(get "wikibase-person" "Marvin Minsky" (:CODE "DEATH-DATE"))
((:yyyymmdd 20160124))
```

### PR DESCRIPTION
**Pull request by chsamlee**
Continuation of #207.

Changes:
- more modular code for extracting date
- people born with a different name and born in/on can be extracted correctly [Bill Clinton]
- words before date, e.g. baptized 17 December 1770 [Beethoven]
- phrases that contains 3+ date-like objects, e.g. 4 BC to AD 30-33 [Jesus]
- location + date in parentheses, nothing else [Augusto La Torre]
- extracting death date outside parentheses [Edward Michael S.]
- `id-info` method in `PersonResolver`: return the parentheses containing birth/death dates without separating them

Future enhancements:
- phrases that indicate a period of time, e.g. early 17th century [Münejjim Bashi]
- adding circas to dates: if we see the date c. 2200 - 2100 BC, it's very likely that 2100 BC is an estimate too

Fixes #77, fixes #87 (note that some date-related infobox queries are also affected by this change), fixes #92, fixes #94, fixes #215 (added test case)

### CODE REVIEW

**Comment by chsamlee** [comment]
Alvaro, can you do a preliminary review of what I have so far?
Also, how do setup Travis so wordnet can be loaded correctly?

File: wikipediabase/date_parser.py [hunk 1]
```diff
@@ -153,36 +106,306 @@ def should_parse_date(date):
     # stage 1: check for certain strings that incidate bad format
     badStrings = ['?', u'\u2013', '/', ' or ']
```
Comment (alvaromorales) [inline wikipediabase/date_parser.py]: Please change camelCase to underscore_separated (`badStrings` -> `bad_strings`); same for all variables below

File: wikipediabase/date_parser.py [hunk 2]
```diff
@@ -153,36 +106,306 @@ def should_parse_date(date):
...
     # TODO in PR 2
     return True
 
 
-def parse_date_yyyymmdd(txt):
+def parse_date_yyyymmdd(val):
     """
     Given a string, returns date in YYYYMMDD format, or None if cannot parse.
+    Parameters
+        txt: <str/tuple> a string that encodes a date, or a (D, M, Y) tuple
```
Comment (alvaromorales) [inline wikipediabase/date_parser.py]: `txt` should be changed to `val` in the docstring

File: wikipediabase/date_parser.py [hunk 3]
```diff
@@ -153,36 +106,306 @@ def should_parse_date(date):
...
+            # context ends with non-alpha character
+            if context in DATE_REFS:
+                return context
+            elif not context[-1].isalpha():
+                return ''
+            # context is a preposition
+            elif context in {'before', 'after', 'by'}:
+                return context
+            elif context in {'in', 'on'}:
+                # I feel that this one may cause us to miss context
```
Comment (alvaromorales) [inline wikipediabase/date_parser.py]: can you elaborate on this comment? what sort of context?

File: wikipediabase/date_parser.py [hunk 4]
```diff
@@ -153,36 +106,306 @@ def should_parse_date(date):
...
+                return ''
+            # context is a verb
+            elif context in {'born', 'died'}:
+                return ''
+            elif (WordNetLemmatizer().lemmatize(context, 'v') != context or
```
Comment (alvaromorales) [inline wikipediabase/date_parser.py]: for performance, you could define `LEMMATIZER = WordNetLemmatizer()` as a global above the class definition, and reference it hear. If the object initialization causes the model files to be re-loaded into memory every time, this might slow things down.

**Comment by alvaromorales** [comment]
Looking good so far! I really like how you modularized the code.

To install WordNet on Travis, add this line to `.travis.yml`: https://github.com/infolab-csail/whoami/blob/master/.travis.yml#L11

Dont' forget to add a test for Marvin Minsky, per #215.

**Comment by chsamlee** [comment]
I think this is ready for full review.

File: wikipediabase/date_parser.py [hunk 5]
```diff
@@ -138,51 +92,373 @@ def should_parse_date(date):
...
+            "baptized 1 January 1111" -> "baptized"
+        Parameters
+            string: <unicode> string that precedes a date-like object
+        Returns
+            (context, type) where
+            context = <unicode> is a phrase that modifies the meaning of date
+            type = 0 if context does not indicate that the date is inexact
+                   1 if context indicates inexact birth/date date
+                   2 if context indicates that date is not birth/death date
+                     and that the real birth/death date extends beyond
```
Comment (alvaromorales) [inline wikipediabase/date_parser.py]: For context types, can you make them into constants defined in `DateUtil`, e.g. `CONTEXT_CERTAIN = 0` or `CONTEXT_INEXACT = 1`. This will allow you to return `('', DateUtil.CONTEXT_CERTAIN)` instead of `('', 0)` (better for readability).

**Comment by alvaromorales** [comment]
Great work!

Can you please add a test case for [Richard Shakespeare](https://en.wikipedia.org/wiki/Richard_Shakespeare)'s birth and death dates? He's like Bill Clinton—a classical "Everything in Wikipedia is Hard" example, and is part of START's test suite.

Are there any dates that are too hard to parse that you decided weren't worth fixing? If so, you should add tests marked as expected failures in case we ever get around to fixing them.

**Comment by chsamlee** [comment]
I am going through another batch of 1000 articles to find hard-to-parse formats since the "irregular" ones I found before are already dealt with. I also found a few bad mistakes that should be fixed before merge: Giovanni Mangone (born towards the end of 15th century -> outputs born on year 15), Reed Painter (fl. 420s–410s BC -> output drops the BC even though it shouldn't), and possibly more as I read through the articles.

**Comment by alvaromorales** [comment]
Looking great! I'd like to see a few more tests for the new `ID-INFO` attribute.

@sfelshin, `ID-INFO` returns an HTML answer with identifier information in the first parens, e.g.,

```
$ telnet host 8023
(get "wikibase-person" "Bill Clinton" (:code "ID-INFO"))
((:html "(born William Jefferson Blythe III, August 19, 1946)"))
```

Note that the value includes the parens. Should the parens be removed by WikipediaBase?

**Comment by sfelshin** [comment]
@chsamlee, please do remove the parens.  Better to keep all knowledge of the specialized formatting in WikipediaBase, so that if Wikipedia changes from parens to something else, we won't have to modify START.

**Comment by alvaromorales** [comment]
lgtm 🚢
